src/lowertest/src/lib.rs

```diff
@@ -0,0 +1,577 @@
+    match to_json(stream_iter, type_name, rti, ctx)? {
+        Some(j) => Ok(Some(serde_json::from_str::<D>(&j).map_err(|e| {
+            format!("String while serializing: {}\nOriginal JSON: {}", e, j)
+        })?)),
+        None => Ok(None),
+    }
+}
+
+/// Deserialize the next `TokenTree` into a `D` object.
+///
+/// See [`to_json`] for the object spec syntax.
+///
+/// `type_name` should be `D` in string form.
+///
+/// `stream_iter` will advance by one `TokenTree` no matter the result.
+pub fn deserialize<D, I, C>(
+    stream_iter: &mut I,
+    type_name: &'static str,
+    rti: &ReflectedTypeInfo,
+    ctx: &mut C,
+) -> Result<D, String>
+where
+    C: TestDeserializeContext,
+    D: DeserializeOwned,
+    I: Iterator<Item = TokenTree>,
+{
+    deserialize_optional(stream_iter, type_name, rti, ctx)?
+        .ok_or_else(|| format!("Empty spec for type {}", type_name))
+}
+
+/// Converts the next `TokenTree` in the stream into deserializable JSON.
+///
+/// Returns `None` if end of stream has been reached.
+///
+/// The JSON string should be deserializable into an object of type
+/// `type_name`.
+///
+/// Default object syntax:
+/// * An enum is represented as `(enum_variant_snake_case <arg1> <arg2> ..)`,
+///   unless it is a unit enum, in which case it can also be represented as
+///   `enum_variant_snake_case`. Enums can have optional arguments, which should
+///   all come at the end.
+/// * A struct is represented as `(<arg1> <arg2> ..)`
+/// * A vec or tuple is represented as `[<elem1> <elem2> ..]`
+/// * None/null is represented as `null`
+/// * true (resp. false) is represented as `true` (resp. `false`)
+/// * Strings are represented as `"something with quotations"`.
+/// * A numeric value like -1 or 1.1 is represented as is.
+/// * You can delimit arguments and elements using whitespace and/or commas.
+///
+/// `ctx` will extend and/or override the default syntax.
+pub fn to_json<I, C>(
+    stream_iter: &mut I,
+    type_name: &str,
+    rti: &ReflectedTypeInfo,
+    ctx: &mut C,
+) -> Result<Option<String>, String>
+where
+    C: TestDeserializeContext,
+    I: Iterator<Item = TokenTree>,
+{
+    // Normalize the type name by stripping whitespace.
+    let type_name = type_name.replace(" ", "");
+    // Eliminate outer `Option<>` and `Box<>` from type names because they are
+    // inconsequential when it comes to creating a correctly deserializable JSON
+    // string.
+    let type_name = if type_name.starts_with("Option<") && type_name.ends_with('>') {
+        &type_name[7..(type_name.len() - 1)]
+    } else if type_name.starts_with("Box<") && type_name.ends_with('>') {
+        &type_name[4..(type_name.len() - 1)]
+    } else {
+        &type_name
+    };
+    if let Some(first_arg) = stream_iter.next() {
+        if let Some(result) =
+            parse_as_enum_or_struct(first_arg.clone(), stream_iter, &type_name, rti, ctx)?
+        {
+            return Ok(Some(result));
+        }
+        if let Some(result) =
```

Comment:
Should we not try override_syntax before any default syntax?

Reply:
`parse_as_enum_or_struct` also calls `override_syntax` (see line 452). I'll add a comment a couple of lines up.

For enums/structs, I wanted to change what gets passed into `override_syntax`. See line 140 in the README.md for the explanation. 

The rationale is that:
The syntax supports specifying a single-arg enum/struct specification as `(<the_arg> )` or `<the_arg>` . I found it really annoying when implementing `override_syntax` for `MirRelationExpr` and `MirScalarExpr` that the code always looks like:
```
match first_arg {
  TokenTree::Ident => {..},
  TokenTree::Literal(literal) => {..},
  TokenTree::Group(group) => {
     match group.stream().into_iter().next() {
         // a repeat of the code for handling first_arg = TokenTree::Ident or Literal
     } 
  }
}
```

If we know the type is supposed to be an enum or a struct, I just want the args to the enum/struct passed to the override and not the potential outer parentheses.